docs: update DESIGN.md document to capture reconciliation refactor (#1900)
opendatahub-io/opendatahub-operator: docs/DESIGN.md

In addition, each ODH component is represented by its own dedicated internal API/custom resource.
Each component CR singleton instance and its sub-resources are managed by a dedicated component controller.
More details on the supported ODH components are provided [here](#component-crs-and-controllers).

The operator also watches [Feature Trackers](#feature-trackers) which are cluster-scoped resources, specifically designed for monitoring and managing objects created via the internal Features API.

Furthermore, there are dedicated [service CRs and controllers](#service-crs-and-controllers) for configuring authentication and monitoring.

Lastly, the ODH operator also manages [accessory controllers](#accessory-controllers) responsible for helper tasks, such as secret generation and cleanup after uninstallation.

A high-level diagram showcasing the majority of the above-mentioned CRs and their reconcilers is provided below.

![Operator Reconciler Architecture](images/odh-operator-reconciliation.png)

### User-facing Custom Resources and their controllers

> The two user-facing CRs are **DSCInitialization** (DSCI) and **DataScienceCluster** (DSC).

Author: Removed the number, added explicit mentions of Auth CR and it being user-facing

Together, these CRs are responsible for configuration of ODH at the platform level.

#### DSCInitialization (DSCI)

> - DSCI CR is created and managed by the ODH operator to perform initial setup that is common for all components.

Reviewer: it is managed but not necessary created by operator.

  - for example, DSCI defines the namespace for ODH component deployments, monitoring config, ServiceMesh config,...

> - Some examples of initial setup include creating namespaces, network policies, SCCs, common configmaps and secrets.

Reviewer: does it create SCC?

> - DSCI instance is a singleton in the k8s/OCP cluster, i.e., only one instance of this CR will always be present in the cluster.

Reviewer: can only say cluster, or OCP , no need k8s/OCP, since we do not support vanilla k8s can cause more confusion

only one instance of this CR will always be present in the cluster. =>
only one instance of this CR can be present in the cluster.

> - DSCI can be deleted to re-run the initial setup without requiring rebuild of the operator.

Reviewer: does this need to be here?

Author: I pretty much kept (and slightly updated) these DSCI and DSC descriptions from the previous version of the document. My thinking was to keep it, as these are the main user-facing CRs and it helped me to understand their purpose

Author: I see, I'll remove it, thanks

- DSCI controller implementation can be found in `internal/controller/dscinitialization` directory.

#### DataScienceCluster (DSC)

- DSC CR is watched by the ODH operator and serves as a single point to configure and enable various ODH components.

> - It is responsible for enabling support for CRDs like Dashboard, Workbenches, Data Science Pipelines, etc.

Reviewer: ```suggestion
- It is responsible for enabling support for CRDs like Dashboard, Workbenches, DataSciencePipelines, etc.
```

> - Like with DSCI, DSC instance is also a singleton in the k8s/OCP cluster. A user can extend/update the CR to enable/disable components.

Reviewer: same here, skip "k8s" and "extend"

- DSC controller implementation can be found in `internal/controller/datasciencecluster` directory.
- Detailed API fields are described in the CRD. Example DSC configurations are provided in the [Examples section](#examples).

### Component CRs and controllers

The ODH operator reconciliation refactor introduced dedicated CRDs for each ODH component.
From there, if enabled, an ODH component is represented by its dedicated singleton CR instance and the owned subresources, all of which are managed by a dedicated component controller. Component CRs are owned by the DSC.
The component controller implementations are located in the subdirectories of `internal/controller/components` directory.
 
List of currently integrated ODH components:

| **ODH Component** | **CR Kind** |
| --- | --- |
| CodeFlare | `CodeFlare` |
| ODH Dashboard | `Dashboard` |
| Data Science Pipelines | `DataSciencePipelines` |
| Feature Store (Feast) | `FeastOperator` |
| KServe | `Kserve` |
| Kueue | `Kueue` |

> | Model Controller | `ModelController` |

Reviewer: this should not be listed

Author: The original document contained the list of supported components. Is this about removing the CR Kind column, or removing the whole table?

Reviewer: only this line "Model Controller " which is for user to interact

| TrustyAI | `TrustyAI` |
| Workbenches (IDEs) | `Workbenches` |

Component reconciliation refactor introduced a generic component reconciler builder and a CLI tool for generating the necessary files, making it easier to integrate new components in the future.
For more details, please refer to the [Component integration guide](https://github.com/opendatahub-io/opendatahub-operator/blob/main/docs/COMPONENT_INTEGRATION.md).

### Feature Trackers

A `FeatureTracker` represents a cluster-scoped resource, specifically designed for monitoring and managing objects created via the internal Features API.
This resource serves a crucial role in cross-namespace resource management, acting as an owner reference for various resources.
The primary purpose of the `FeatureTracker` is to enable efficient garbage collection by Kubernetes.
This is essential for ensuring that resources are automatically cleaned up and reclaimed when they are no longer required.

### Service CRs and controllers

In addition to the previously-mentioned CRs and their controller, the operator also manages the following ODH service CRs:
- `Auth`
  - responsible for auth configuration

>   - singleton instance in the k8s/OCP cluster

Reviewer: same here "k8s"

  - controller implementation located in `internal/controller/services/auth` 
- `Monitoring`
  - responsible for monitoring configuration

>   - singleton instance in the k8s/OCP cluster

Reviewer: same here "k8s"

  - controller implementation located in `internal/controller/services/monitoring`

### Accessory controllers

The ODH operator currently manages several accessory controllers that handle parts of its functionality.
The currently used accessory controllers are listed below:
- Cert ConfigMap Generator controller
  - responsible for generating the ConfigMap with certificates (`odh-trusted-ca-bundle`), which includes cluster-wide trusted-ca bundle and custom ca bundle in every new namespace created.
  - controller implementation located in `internal/controller/services/certconfigmapgenerator`.
- Secret Generator controller
  - responsible for generating Secret used by the ODH Dashboard.
  - controller implementation located in `internal/controller/services/secretgenerator`.
- ServiceMesh controller

>   - responsible for configuring/removing ServiceMesh.

Reviewer: ```suggestion
  - responsible for configuring ServiceMesh v2 and Authorino related resources.
```

  - controller implementation located in `internal/controller/services/servicemesh`.
- Setup controller
  - responsible for managing the ConfigMap that triggers the cleanup/uninstallation of ODH.
  - handles the cleanup logic itself.
  - controller implementation located in `internal/controller/services/setup`.

## Examples

1. Enable all components in DataScienceCluster

```console
apiVersion: datasciencecluster.opendatahub.io/v1
kind: DataScienceCluster
metadata:
  name: example
spec:
  components:
    codeflare:
      managementState: Managed
    dashboard:
      managementState: Managed
    datasciencepipelines:
      managementState: Managed

>     kserve:
>       managementState: Managed
>       serving:
>         ingressGateway:
>           certificate:
>             type: OpenshiftDefaultIngress
>         managementState: Managed
>         name: knative-serving

Reviewer: ```suggestion
    kserve:
      nim:
        managementState: Managed
      rawDeploymentServiceConfig: Headless
      serving:
        ingressGateway:
          certificate:
            type: OpenshiftDefaultIngress
        managementState: Managed
        name: knative-serving
```
